plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py

```diff
@@ -0,0 +1,180 @@
+DEFAULT_UP_SECONDS = 10 * HOURS_TO_SECONDS  # 10 hours
+
+
+def execute_command(cmd):
+    """
+    Execute a command in the shell.
+    """
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    logger.info(f"cmd: {cmd}")
+    stdout, stderr = process.communicate()
+    if process.returncode != 0:
+        raise RuntimeError(f"Command {cmd} failed with error: {stderr}")
+    logger.info(f"stdout: {stdout}")
+    logger.info(f"stderr: {stderr}")
+
+
+def download_file(url, target_dir="."):
+    """
+    Download a file from a given URL using fsspec.
+
+    Parameters:
+    - url (str): The URL of the file to download.
+    - target_dir (str, optional): The directory where the file should be saved. Defaults to current directory.
+
+    Returns:
+    - str: The path to the downloaded file.
+    """
+
+    if not url.startswith("http"):
+        raise ValueError(f"URL {url} is not valid. Only http/https is supported.")
+
+    # Derive the local filename from the URL
+    local_file_name = os.path.join(target_dir, os.path.basename(url))
+
+    fs = fsspec.filesystem("http")
+
+    # Use fsspec to get the remote file and save it locally
+    logger.info(f"Downloading {url}... to {os.path.abspath(local_file_name)}")
+    fs.get(url, local_file_name)
+    logger.info("File downloaded successfully!")
+
+    return local_file_name
+
+
+def download_vscode(
+    code_server_remote_path: str,
+    code_server_dir_name: str,
+):
+    """
+    Download vscode server and plugins from remote to local and add the directory of binary executable to $PATH.
+
+    Parameters:
+    - code_server_remote_path (str): The URL of the code-server tarball.
+    - code_server_dir_name (str): The name of the code-server directory.
+    Return:
+    - str: The path to the code-server binary.
+    """
+
+    # If the code server already exists in the container, skip downloading
+    executable_path = shutil.which(EXECUTABLE_NAME)
+    if executable_path is not None:
+        logger.info(f"Code server binary already exists at {executable_path}")
+        logger.info("Skipping downloading code server...")
+        return
+
+    logger.info("Code server is not in $PATH, start downloading code server...")
+
+    # Create DOWNLOAD_DIR if not exist
+    logger.info(f"DOWNLOAD_DIR: {DOWNLOAD_DIR}")
+    os.makedirs(DOWNLOAD_DIR, exist_ok=True)
+
+    logger.info(f"Start downloading files to {DOWNLOAD_DIR}")
+
+    # Download remote file to local
+    code_server_tar_path = download_file(code_server_remote_path, DOWNLOAD_DIR)
+
+    # Extract the tarball
+    with tarfile.open(code_server_tar_path, "r:gz") as tar:
+        tar.extractall(path=DOWNLOAD_DIR)
+
+    code_server_dir_path = os.path.join(DOWNLOAD_DIR, code_server_dir_name)
+
+    code_server_bin_dir = os.path.join(code_server_dir_path, "bin")
+
+    # Add the directory of code-server binary to $PATH
+    os.environ["PATH"] = code_server_bin_dir + os.pathsep + os.environ["PATH"]
```

Comment:
is `os.join()` better than manually using `+`?

Reply:
Do you mean os.path.join()? In this case, I think we cannot use os.path.join() because we are going to concatenate $PATH with new path by ":" (os.pathsep) instead of combining two path to one path. For example, if $PATH is /usr/bin:/local/bin and code_server_bin_dir is /tmp/bin, then the result will be /tmp/bin:/usr/bin:/local/bin.

Reply:
good point!